OpenSim/Auxiliary/auxiliaryTestFunctions.h

```diff
@@ -209,17 +209,116 @@ OpenSim::Object* randomize(OpenSim::Object* obj)
-    else
-      fileNew << line << "\n";
-  }
+    const std::string& filenameNew) {
+    std::regex versionline{ R"([ \t]*version[ \t]*=[ \t]*\d[ \t]*)" };
+    std::ifstream fileOld{ filenameOld };
+    std::ofstream fileNew{ filenameNew };
+    std::string line{};
+    while (std::getline(fileOld, line)) {
+        if (std::regex_match(line, versionline))
+            fileNew << "version=1\n";
+        else
+            fileNew << line << "\n";
+    }
+}
+
+// Estimate the memory usage of a *creator* that heap allocates an object
+// of type C and returns a pointer to it. Creator can also perform any 
+// initialization before returning the pointer.
+template <typename C, typename T>
+size_t estimateMemoryChangeForCreator(T creator, const size_t nSamples = 100)
+{
+    std::vector<std::unique_ptr<C>> pointers;
+    std::vector<size_t> deltas;
+
+    for (size_t i = 0; i < nSamples; ++i) {
+        size_t mem0 = getCurrentRSS();
+        // Execute the desired creator 
+        // store in unique_ptrs to delay deletion
+        pointers.push_back(std::unique_ptr<C>(creator()));
+        // poll the change in memory usage
+        size_t mem1 = getCurrentRSS();
+        // change in memory usage (negative values are invalid)
+        size_t delta = mem1 > mem0 ? mem1 - mem0 : 0;
+        if(delta) // store only valid values (creator cannot create 0 bytes)
+            deltas.push_back(delta);
+    }
+
+    OPENSIM_THROW_IF(deltas.size() < 2, OpenSim::Exception,
+        "Insufficient number of nonzero samples to estimate memory change. "
+        "Consider increasing the number of samples.");
+
+    size_t nmedian = deltas.size() / 2;
+    // sort the deltas up to and including the nth element
+    std::nth_element(deltas.begin(), deltas.begin() + nmedian, deltas.end());
+
+    return deltas[nmedian];
+}
+
+// Determine if getRSS is providing reliable estimates of memory usage by
+// testing against an allocation of known size and verifying that the change
+// memory use is detected. Employ this method to validate the use of
+// memory use estimators (e.g. estimateMemoryChangeForCreator and
+// estimateMemoryChangeForCommand). Do this at the beginning of your test 
+// involving checks for memory use.
+void validateMemoryUseEstimates()
+{
+    size_t nSamples = 5;
+    // Approximate size of a small OpenSim model
+    size_t size = 1000 * 1024; // 1K * 1KB = 1MB;
+
+    struct Block {
+        Block(size_t size) {
+            // allocate block of stuff of specified size
+            p = (char*)malloc(size);
+            // do some random initialization
+            for (size_t i = 0; i < size; i+=1024) {
+                p[i] = rand();
+            }
+        }
+        ~Block() {
+            free(p);
+        }
+        // member is pointer to allocated block
+        char* p{};
+    };
+
+    auto creator = [size]() { 
+        return new Block(size);
+    };
+
+    size_t delta = estimateMemoryChangeForCreator<Block>(creator, nSamples);
```

Comment:
I suggest enclosing this line in a try/catch and re-throwing the exception here so the message indicates whether the "Insufficient number of nonzero samples to estimate memory change" exception occurred pre- or post-"validation".